services/registration/src/common/adjustScores.ts

```diff
@@ -0,0 +1,176 @@
+ * Calculates line of best fit given an array of points.
+ * @param points the array of points
+ * @returns line of best fit function
+ */
+const computeLineOfBestFit = (points: [number, number][]): ((x: number) => number) => {
+  let avgx = 0,
+    avgy = 0,
+    slopeNumerator = 0,
+    slopeDenominator = 0,
+    slope = 0,
+    b = 0;
+
+  points.forEach(([x, y]) => {
+    avgx += x;
+    avgy += y;
+  });
+
+  avgx /= points.length;
+  avgy /= points.length;
+
+  points.forEach(([x, y]) => {
+    slopeNumerator += (x - avgx) * (y - avgy);
+    slopeDenominator += (x - avgx) * (x - avgx);
+  });
+
+  if (slopeDenominator === 0) {
+    console.warn("Unexpected condition: slope denominator is zero... returning identity map");
+    return x => x;
+  }
+
+  slope = slopeNumerator / slopeDenominator;
+  b = avgy - slope * avgx;
+
+  return x => slope * x + b;
+};
+
+/**
+ * Takes in an array of points, and uses a line of best fit to create a mapping
+ * @param points the array of points
+ * @param from minimum value of the mapping
+ * @param to maximum value of the mapping
+ * @returns interpolated mapping of points as an object
+ */
+const interpolateValuePairs = (points: [number, number][]) => {
+  const line = computeLineOfBestFit(points);
+  const adjustedMap: Map<string, number> = new Map<string, number>();
+
+  for (let i: number = MIN_GRADING_SCORE; i <= MAX_GRADING_SCORE; i++) {
+    let val = Number(line(i).toFixed(2));
+    if (val > MAX_GRADING_SCORE) {
+      val = MAX_GRADING_SCORE;
+    } else if (val < MIN_GRADING_SCORE) {
+      val = MIN_GRADING_SCORE;
+    }
+    val = Math.round(val * 2) / 2;
+    adjustedMap.set(i.toString(), val);
+  }
+
+  // In case the line of best fit is downward sloping, we need to flip the mapping
+  if (
+    adjustedMap.get(MAX_GRADING_SCORE.toString())! < adjustedMap.get(MIN_GRADING_SCORE.toString())!
+  ) {
+    const flippedMap: Map<string, number> = new Map<string, number>();
+    let start = MIN_GRADING_SCORE;
+    let end = MAX_GRADING_SCORE;
+    while (start <= end) {
+      flippedMap.set(start.toString(), adjustedMap.get(end.toString())!);
+      flippedMap.set(end.toString(), adjustedMap.get(start.toString())!);
+      start++;
+      end--;
+    }
+    return flippedMap;
+  }
+
+  return adjustedMap;
+};
+
+/**
+ * Given two arrays of criteria scores, computes the calibration mapping.
+ *
+ * @param graderCriteriaScores
+ * @param groundTrustCriteriaScores
+ * @returns calibration mapping
+ */
+
+const computeCalibrationMapping = (
+  graderCriteriaScores: CriteriaScores,
+  groundTrustCriteriaScores: CriteriaScores
+) => {
+  if (graderCriteriaScores.length !== groundTrustCriteriaScores.length) {
+    throw new Error("Number of grader scores does not match number of ground truth scores.");
+  }
+
+  const allCriterias = new Set<string>();
+  Object.keys(rubricMapping).forEach(criteria => allCriterias.add(criteria));
+
+  const calibrationMapping = Array.from(allCriterias).map(criteria => {
```

Comment:
Can delete the `allCriterias` variable and just do `Object.keys(rubricMapping)` since we shouldn't have duplicate keys.